corgidrp/detector.py

```diff
@@ -520,7 +528,149 @@ def calc_sat_fwc(emgain_arr,fwcpp_arr,fwcem_arr,sat_thresh):
+        	percens_resel (np.array): percentage error of the flat field
+    """
+    
+    n = 420
+    full_qe=np.random.normal(1,.03,(n,n))
+    full_residuals = np.zeros((n,n))
+    rad1 = 1.26; rad4 = 1.75
+    aperture1 = CircularAperture((np.ceil(rad1), np.ceil(rad1)), r=rad1)
+    aperture4 = CircularAperture((np.ceil(rad4), np.ceil(rad4)), r=rad4)
+    mask1= aperture1.to_mask().data
+    mask4= aperture4.to_mask().data
+    if planet=='neptune':
+        rad = 50
+    elif planet=='uranus':
+        rad=65
+
+    for i in np.arange(len(residual_images)):
+        nx = np.arange(0,residual_images[i].shape[1])
+        ny = np.arange(0,residual_images[i].shape[0])
+        nxx,nyy = np.meshgrid(nx,ny)
+        nrr = np.sqrt((nxx-rad-5)**2 + (nyy-rad-5)**2)
+        nrr_copy = nrr.copy()
+        nrr_copy[nrr<rad] = residual_images[i][nrr<rad] 
+        nrr_copy[nrr>=rad] = None 
+        ymin = int(cent[i][0])
+        ymax = int(cent[i][1])
+        xmin = int(cent[i][2])
+        xmax = int(cent[i][3])
+
+        bool_innotzero = np.logical_and(nrr<rad,full_residuals[ymin:ymax,xmin:xmax]!=0)
+        bool_iniszero = np.logical_and(nrr<rad,full_residuals[ymin:ymax,xmin:xmax]==0)
+        bool_outisnotzero = np.logical_and(nrr>=rad,full_residuals[ymin:ymax,xmin:xmax]!=0)
+        
+        full_residuals[ymin:ymax,xmin:xmax][bool_innotzero] = (nrr_copy[bool_innotzero] + full_residuals[ymin:ymax,xmin:xmax][bool_innotzero])/2
+        full_residuals[ymin:ymax,xmin:xmax][bool_iniszero] = nrr_copy[bool_iniszero]
+        full_residuals[ymin:ymax,xmin:xmax][bool_outisnotzero] = full_residuals[ymin:ymax,xmin:xmax][bool_outisnotzero]
+
+    if band == '4':
+        full_qe_resel = ndimage.convolve(full_qe, mask4)
+        full_residuals_resel = ndimage.convolve(full_residuals,mask4)
+        percens_resel = (((full_qe_resel - full_residuals_resel)/full_qe_resel)*100)
+    
+    elif band== '1':
+        full_qe_resel = ndimage.convolve(full_qe, mask1)
+        full_residuals_resel = ndimage.convolve(full_residuals,mask1)
+        percens_resel = (((full_qe_resel - full_residuals_resel)/full_qe_resel)*100)
+    
+    full_residuals[full_residuals==0] = None
+    nx = np.arange(0,percens_resel.shape[1])
+    ny = np.arange(0,percens_resel.shape[0])
+    nxx,nyy = np.meshgrid(ny,nx)
+    nrr = np.sqrt((nxx-n/2)**2 + (nyy-n/2)**2)
+    full_residuals[nrr>165]= 1
+    percens_resel[nrr>165] = 0
+    full_qe[nrr>165]= 1
+    full_qe=np.pad(full_qe, ((302,302),(302,302)), mode='constant',constant_values=(1))
+    full_residuals=np.pad(full_residuals, ((302,302),(302,302)), mode='constant',constant_values=(1))
+    percens_resel=np.pad(percens_resel, ((302,302),(302,302)), mode='constant',constant_values=(1))
+    
+    return (full_qe,full_residuals,percens_resel)
+    
+    
+def create_onsky_flatfield(dataset, planet=None,band=None):
+    """Turn this dataset of raster scanned image frames of neptune or uranus into on-sky calibrated flat field 
```

Comment:
@ramya-anche - can you please also add more detail for this function? in particular, please add the expectations for the input products - what data processing do we expect the images to have already gone through?

flat fielding is part of the data processing of L2a to L2b - and it is essentially the last step of the L2b, so you should write that this function assumes that the input data is a L2b image just prior to flat fielding, which is an image that has been "Divide by Kgain, divide by EM Gain, subtract Dark, desmear"